libklog/libklog.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Phoenix-RTOS klog driver
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <paths.h>
+#include <string.h>
+#include <sys/msg.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <posix/utils.h>
+
+#include "libklog.h"
+
+static struct {
+	char __attribute__((aligned(8))) stack[2048];
+	libklog_write_t ttywrite;
+	void *ttyctx;
+} libklog_common;
+
+
+extern int sys_open(const char *filename, int oflag, ...);
+
+
+static void pumpthr(void *arg)
+{
+	char buf[256];
+	int fd, ret;
+	oid_t dev;
+	char *name;
+
+	dev.id = 0;
+	dev.port = 0;
+
+	if ((fd = open(_PATH_KLOG, O_RDONLY)) < 0) {
+		/* On some architectures devFS might not be bound
+		 * to /dev directory yet, which makes /dev/kmsg path not resolvable.
+		 * To make devfs/kmsg resolvable, we need to register it first.
+		 */
+		strcpy(buf, "devfs");
+		if ((name = strrchr(_PATH_KLOG, '/')) == NULL)
+			endthread();
+
+		strcat(buf, name);
+
+		if (portRegister(0, buf, &dev) != 0)
+			endthread();
+
+		/* open() treats paths not starting with '/' slash as local */
+		if ((fd = sys_open(buf, O_RDONLY, 0)) < 0)
+			endthread();
+	}
+
+	while (1) {
+		ret = read(fd, buf, sizeof(buf));
+		if (ret <= 0) {
+			if (errno == -EINTR)
+				continue;
+			else
+				break;
+		}
+		libklog_common.ttywrite(libklog_common.ttyctx, buf, ret);
+	}
+	endthread();
+}
+
+
+int libklog_init(libklog_write_t clbk, void *ctx)
+{
+	oid_t dev;
+	int err;
+
+	libklog_common.ttywrite = clbk;
+	libklog_common.ttyctx = ctx;
```

Comment:
We don't really need this ctx - user of the lib can prepare callback that handles writes as intended.
Pros:
- we don't violate MISRA rule about pointer casting
Cons:
- user of the lib need to prepare specific handler (you do this anyway in imx6ull-uart ;))